From 6d0cdfa97c90b9f2eaa0e4915cdc1f7caa4bfebe Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sun, 8 Sep 2019 22:36:24 +0200
Subject: [PATCH 24/53] bmq-5.2: unbreak systemd-detect-virt

/proc/<pid>/sched shows nothing, thus systemd-detect-virt always assumes
the system runs inside a container.

Modify proc_sched_show_task() to show the very first line, so systemd is
happy.

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 kernel/sched/bmq.c | 17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index b1cc2518157f..c406bfbcfdeb 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -54,6 +54,18 @@
 /* Reschedule if less than this many Î¼s left */
 #define RESCHED_NS		(100 * 1000)
 
+/*
+ * This allows printing both to /proc/sched_debug and
+ * to the console
+ */
+#define SEQ_printf(m, x...)			\
+ do {						\
+	if (m)					\
+		seq_printf(m, x);		\
+	else					\
+		pr_cont(x);			\
+ } while (0)
+
 static inline void print_scheduler_version(void)
 {
 	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.99 by Alfred Chen.\n");
@@ -5913,7 +5925,10 @@ void ia64_set_curr_task(int cpu, struct task_struct *p)
 #ifdef CONFIG_SCHED_DEBUG
 void proc_sched_show_task(struct task_struct *p, struct pid_namespace *ns,
 			  struct seq_file *m)
-{}
+{
+	SEQ_printf(m, "%s (%d, #threads: %d)\n", p->comm, task_pid_nr_ns(p, ns),
+						get_nr_threads(p));
+}
 
 void proc_sched_set_task(struct task_struct *p)
 {}
-- 
2.24.0.155.gd9f6f3b619

